### PR TITLE
Rng kernels

### PR DESCRIPTION
**Pull request by fcharras**
Same PR than https://github.com/soda-inria/sklearn-numba-dpex/pull/42 that I closed by mistake

This time, with tests

The tests take randomgen implementation as reference, but I've found errors in randomgen while writing the tests, until the errors are confirmed (reported in https://github.com/bashtage/randomgen/issues/321 ) our implementation is corrupted with the same errors for the moment. 

Maybe another reference implementation would be preferable if you have suggestions. I prefer not to use `numba.cuda.random` as reference because it requires a GPU to check it.

### CODE REVIEW

**Comment by fcharras** [comment]
@jjerphan to answer your suggestion in the other PR, I'd rather check that the RNG returns the exact same values than with a reliable reference implementation, than independently trying to test the good properties of randomness of the generator, which sounds harder.

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:41]: Can you add docstrings to specify the kind of pseudo-random value this is generating (I assume float32 in [0, 1).)?

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:45]: Some comment regarding docstrings.

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:205]: We can easily come up with Python interface using Cython if needed. Let me know if this would be useful and I can work on this.

**Review by jjerphan — COMMENTED** [review]
Some comments from a code and maintenance perspectives after a first walk in this bit fields' farm.

I need to get into the theoretical aspects first before continuing reviews.

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:54]: For consistency based on what follows.
```suggestion
        convert_const = float64(uint64(1) << uint32(53))
```

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:60]: Ditto.
```suggestion
            return float64(x >> convert_rshift) * (convert_const_one / convert_const)
```

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:232]: It looks like this part of the comment is duplicated and can be removed.

```suggestion
```

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:298]: Is this apparently erroneous as well?

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:301]: To me, this naming is clearer.
```suggestion
64_as_uint32 = uint32(64)
```

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:307]: If this rotation only works for numbers encoded on 32 bits, I think this must be documented.

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/random.py:98]:
```suggestion
    This initializes the RNG states so that states in the array corresponds
    to subsequences the separated by 2**64 steps from each other in the main
    sequence.
    Therefore, as long as no thread requests more than 2**64 random numbers,
    all the RNG states produced by this function are guaranteed to be independent.
```

**Review by jjerphan — COMMENTED** [review]
Another pass for a few last comments.

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:307]: It works for 64 bits number, it's the shift amount that can be encoded as a `uint32` (it's an int ranging from 0 to 64)

IDK if there's a particular reason to force the `uint32` type for the shift amount but I'm refering to the same instructions in `numba.cuda.random`.

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:298]: Yes, this block is required by the erroneous block.

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:205]: There's not too much complexity in this code but that can give us more confidence in my copy pasting skills :grin:

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:60]: That's better for readability, but per casting rules we already have a `float64` and the comment already states that a `float64` is returned. This formula is used by all reference implementations so I'd also be afraid to add mistakes :-/

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [inline sklearn_numba_dpex/common/random.py:232]: inline comment that hint at the difference there (xoroshirop vs xoroshiropp)

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py]: add docstring that document x dtype

**Comment by fcharras** [comment]
LGTM ?

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (jjerphan) [inline sklearn_numba_dpex/common/tests/test_random.py:146]: Should we preferably use fixtures from scikit-learn for tests assertions (i.e. [`sklearn.utils._testing.assert*`](https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/_testing.py))?

**Review by jjerphan — APPROVED** [review]
LGTM. Just one comment regarding assertions.

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_random.py:146]: I didn't want to add a dependency on sklearn but that's dumb sklearn is already a mandatory dependency :facepalm: will fix & merge

File: sklearn_numba_dpex/common/random.py
Comment (jjerphan) [reply sklearn_numba_dpex/common/random.py]: This was not entirely identical: `_xoroshiro128p_next` is called in this commented code, whereas `_xoroshiro128pp_next` is called, as indicated by @fcharras in a comment which was integrated in the meantime.

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (jjerphan) [reply sklearn_numba_dpex/common/tests/test_random.py:146]: Do not :facepalm: yourself: it's fine. :slightly_smiling_face:

Comment (ogrisel) [reply sklearn_numba_dpex/common/tests/test_random.py:146]: I am reviewing this now. Please wait for my review before merging.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:106]: Could you please explain how to chose this value? The tests only use `n_states=1` if I am not mistaken.

I guess it has to do with the amount of parallelism you want to allow when generating a large number of random numbers?

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:131]:
```suggestion
        Use SplitMix64 to generate an xoroshiro128p state from a uint64 seed.
```

**Review by ogrisel — COMMENTED** [review]
I find the tests in this PR quite unexpected, because they only ever generate one random number per kernel invocation which is really not the canonical usage for a GPU RNG implementation.

Could you please add a test that generate `1e7` values in parallel (with a typical value for `n_states`, not sure how to choose this depending on the device) and check that all the returned values are between `[0, 1)` and that binning in 10 bins using something like

```python
np.histogram(np.random.rand(int(1e7)), bins=np.linspace(0, 1, 11))[0]
```

returns around 1e6 (+/- 1e3) values per bin for a few arbitrary seeds and values for `n_states`?

This test should be parametrized both for `dtype=npfloat32` and `np.float64`.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:78]: PEP257 nitpick.

```suggestion
            """Convert uint64 to float32 value in the range [0.0, 1.0)

            NB: this is different than original numba.cuda.random code. Instead of
            generating a float64 random number before casting it to float32, a float32
            number is generated from uint64 without intermediate float64. This enables
            compatibility with devices that do not support float64 numbers.
            However is seems to be exactly equivalent e.g it passes the float precision
            test in sklearn.
            """
```

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:106]: > I guess it has to do with the amount of parallelism you want to allow when generating a large number of random numbers?

yes. n_states = n_threads

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_random.py:47]: > we should get the same 10 values, no?

we should, let's test this

**Comment by fcharras** [comment]
> I find the tests in this PR quite unexpected

I wanted to avoid the compilcated test on rng quality and associated maintain burden, and only test that we have the same numbers than a reference implementation (here the implementation provided in `randomgen`). The first few numbers starting with the same seed should be enough to prove that the rng is rigorously identical to the reference ?

(there's a weakness in this plan with the issue on randomgen _jump_ function but at least it holds for the _next_ function)

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [reply sklearn_numba_dpex/common/random.py:106]: Please make that explicit in the docstring for this parameter.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [reply sklearn_numba_dpex/common/random.py:29]: Ok I did not know that. Let's keep this name and improve the docstring then.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:30]:
```suggestion
    """Returns one `uint64` random integer
    
    Similar to numpy.random.BitGenerator.random_raw but with parallelism
    controlled by the number of states.
    """
```

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:37]:
```suggestion
def make_random_raw_kernel():
    """Returns random `uint64` integers
    
    Similar to numpy.random.BitGenerator.random_raw but with parallelism
    controlled by the number of states.
    """
```

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:106]: The number of threads [is 2**64 ](https://prng.di.unimi.it/)so we should be fine :grin: 

image

**Comment by fcharras** [comment]
I pushed the suggestions / answered the comments, will work on the finer test for parallel rng tomorrow.

**Comment by fcharras** [comment]
doesn't 1e7 sound a lot for a unitt test ran in a github action runner ? maybe there's a smaller setup that is still enough ?

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [reply sklearn_numba_dpex/common/random.py]: > and really wouldn't work with parallelism. It's not fit for parallel calls...

Yes I realized that after posting my comment and updated the suggestion to only refer to the numpy name.

**Comment by ogrisel** [comment]
> doesn't 1e7 sound a lot for a unitt test ran in a github action runner ? maybe there's a smaller setup that is still enough ?

As you wish. Maybe 1e6 is enough and you can reduce the number of bins, e.g. to `np.linspace(0, 1, 6)` to get 2e5 +/- 1e3 in each bin or something.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:199]:
```suggestion
    def _xoroshiro128pp_jump(states, state_idx):
        """Advance the RNG in ``states[state_idx]`` by 2**64 steps."""
        s0 = jump_init
        s1 = jump_init

        for i in range(long_2):
            if i == zero_idx:
                jump_const = jump_const_1
            else:
                jump_const = jump_const_2
            for b in range(long_64):
                if jump_const & jump_const_3 << uint32(b):
                    s0 ^= states[state_idx, zero_idx]
                    s1 ^= states[state_idx, one_idx]
                _xoroshiro128pp_next(states, state_idx)

        states[state_idx, zero_idx] = s0
        states[state_idx, one_idx] = s1
```

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:303]:
```suggestion
@dpex.func
def _xoroshiro128pp_next(states, state_idx):
    """Return the next random uint64 and advance the RNG in states[index]."""
    s0 = states[state_idx, zero_idx]
    s1 = states[state_idx, one_idx]
    result = _rotl(s0 + s1, next_rot_1) + s0

    s1 ^= s0
    states[state_idx, zero_idx] = _rotl(s0, next_rot_2) ^ s1 ^ (s1 << shift_1)
    states[state_idx, one_idx] = _rotl(s1, next_rot_3)

    return result
```

**Review by ogrisel — COMMENTED** [review]
More feedback based on my updated understanding.

File: sklearn_numba_dpex/common/tests/test_kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_kernels.py:106]: github is bugged because of the conflict after the merge of the reduction kernels, the diff will display properly when main is pulled. (I've opened a lot of cascades of PRs that rely on each other to make review and step-by-step merge easier)

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py:109]: About the scope of this PR the purpose here is not to provide an high-level interface dedicated to user that just generate random number like e.g `numpy.random.rand` would do. For me it was only about serving the need of the  kmeans++ where the device functions (those defined with `@dpex.func`) are used in to have rng in each thread fused with other instructions. It is a lower level use where it is up to the user to choose carefully `n_states` and then use it accordingly in other kernels. And we're not preventing bad use, like having several threads using the same state. Maybe I should have made it clearer in the description of the PR so you can start with a better understanding a priori.

(also, there is rng with better user interface, e.g in jax each thread use a different hash function on integers)

That being said I think your docstring is fine and we can even omit to document the constraint of not having more than `2**64` threads iterate on more than `2**64` values since those are virtually impossible to reach ?

> You mean the maximum number of threads that we can use safely?
yes, the algorithm is safe as long as `n_states < 2**64` and each state not updated more than `2**64` times.

File: sklearn_numba_dpex/common/tests/test_kernels.py
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_kernels.py]: Actually not bugged, the cascade of PR was set such that reduction kernels was merged on this branch. Fixed that and merged them to master.

**Comment by fcharras** [comment]
Had some troubles writing the rng quality test but here it is + applied all other suggestions. TY for the docstrings :smile:

File: sklearn_numba_dpex/common/random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/random.py]: Actually restoring it, a test rightly uses it (to test the jump function).

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (ogrisel) [reply sklearn_numba_dpex/common/tests/test_random.py:52]: This comment has not yet been addressed I think.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [reply sklearn_numba_dpex/common/random.py:42]: This suggestion was not considered yet.

File: sklearn_numba_dpex/common/random.py
Comment (ogrisel) [inline sklearn_numba_dpex/common/random.py:59]:
```suggestion
    The returned kernel function takes two arguments:
```

**Review by ogrisel — APPROVED** [review]
There are a few suggestions left in the diff view of the PR:

https://github.com/soda-inria/sklearn-numba-dpex/pull/45/files

pending those. LGTM!

File: sklearn_numba_dpex/common/tests/test_random.py
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_random.py:52]: I added it in the already existing unittest

**Pull request merged**
